checker/bin/infer-and-annotate.sh

```diff
@@ -0,0 +1,138 @@
+# received as arguments.
+# TODO: Handle the following limitation: This function makes the assumption
+# that every argument starts with a hyphen. It means one cannot pass arguments
+# such as -processorpath and -source, which are followed by a value.
+read_input() {
+    processor=$1
+    cp=$2:$AFU_JAR
+    # Ignores first two arguments (processor and cp).
+    shift
+    shift
+    extra_args=""
+    java_files=""
+    jaif_files=""
+    for i in "$@"
+    do
+        # This function makes the assumption that every extra argument
+        # starts with a hyphen. The rest are .java files.
+        case "$1" in
+            -*)
+                extra_args="$extra_args $1"
+            ;;
+            *.jaif)
+                jaif_files="$jaif_files $1"
+            ;;
+            *.java)
+                java_files="$java_files $1"
+            ;;
+        esac
+        shift
+    done
+}
+
+# Iteratively runs the Checker
+infer_and_annotate() {
+    mkdir -p $TEMP_DIR
+    DIFF_JAIF=firstdiff
+    # Create/clean signature-inference directory.
+    rm -rf $SIGNATURE_INFERENCE_DIR
+    mkdir -p $SIGNATURE_INFERENCE_DIR
+    # If there are .jaif files as input, copy them.
+    for file in $jaif_files;
+    do
+        cp $file $SIGNATURE_INFERENCE_DIR/
+    done
+
+    # Perform inference and add annotations from .jaif to .java files until
+    # $PREV_ITERATION_DIR has the same contents as $SIGNATURE_INFERENCE_DIR.
+    while [ "$DIFF_JAIF" != "" ]
+    do
+        # Updates $PREV_ITERATION_DIR folder
+        rm -rf $PREV_ITERATION_DIR
+        cp -r $SIGNATURE_INFERENCE_DIR $PREV_ITERATION_DIR
+
+        # Runs CF's javac
+        ${CHECKERFRAMEWORK}/checker/bin/javac -d $TEMP_DIR/ -cp $cp -processor $processor -AinferSignatures $extra_args $java_files || true
+        # Deletes .unannotated backup files. This is necessary otherwise the
+        # insert-annotations-to-source tool will use this file instead of the
+        # updated .java one.
+        for file in $java_files;
+        do
+            rm -f "${file}.unannotated"
+        done
+        if [ ! `find $SIGNATURE_INFERENCE_DIR -prune -empty` ]
+        then
+            # Only insert annotations if there is at least one .jaif file.
+            insert-annotations-to-source -i $SIGNATURE_INFERENCE_DIR/* $java_files
+        fi
+        # Updates DIFF_JAIF variable.
+        # diff returns exit-value 1 when there are differences between files.
+        # When this happens, this script halts due to the "set -e"
+        # in its header. To avoid this problem, we add the "|| true" below.
+        DIFF_JAIF="$(diff -qr $PREV_ITERATION_DIR $SIGNATURE_INFERENCE_DIR || true)"
+    done
+    clean
+}
+
+clean() {
+    # It might be good to keep the final .jaif files.
+    # rm -rf $SIGNATURE_INFERENCE_DIR
+    rm -rf $PREV_ITERATION_DIR
+    rm -rf $TEMP_DIR
+    for file in $java_files;
+        do
+            rm -f "${file}.unannotated"
```

Comment:
If the user has previously run insert-annotations-to-source on their code (for a different reason), they could lose their only copy of their unannotated code. You might want to document this behavior.